### PR TITLE
fix(deps): bump internal heddle-grpc consumers to ^0.3 + asserter check (heddle#76)

### LINKED ISSUE #76: Bump internal workspace heddle-grpc deps to ^0.3 (unblocks auto-publish)
## Context

heddle#63 r2 bumped heddle-grpc from 0.2.5 → 0.3.0 (breaking RPC rename, semver-pre-1 minor bump). The publish workflow ([heddle#72](https://github.com/HeddleCo/heddle/issues/72)) auto-fired on the merge of heddle#70 — and **correctly failed loud** because internal workspace crates still pin \`heddle-grpc = "^0.2"\`:

\`\`\`
error: failed to select a version for the requirement \`heddle-grpc = "^0.2"\`
candidate versions found which didn't match: 0.3.0
required by package \`heddle-cli v0.2.4\`
\`\`\`

The fix is mechanical: bump 3 internal Cargo.toml files. Each currently pins \`version = "0.2"\`:

- \`crates/daemon/Cargo.toml:18\`
- \`crates/cli/Cargo.toml:48\`
- \`crates/client/Cargo.toml:51\`

All need: \`version = "0.3"\`.

## Why this matters

- The auto-publish workflow is currently red on every main merge (it tries to publish heddle-grpc 0.3.0, hits the version-mismatch resolution failure, exits 1). This means **no future publishes can happen until this fixes**.
- heddle-grpc 0.3.0 is NOT on crates.io yet — so weft#139 and tapestry#57 (the downstream PRs that depend on it) cannot merge until this lands + heddle-grpc 0.3.0 publishes.

## Acceptance criteria

- [ ] Bump \`heddle-grpc\` version requirement in the 3 internal Cargo.toml files from \`"0.2"\` to \`"0.3"\`.
- [ ] Confirm \`cargo metadata --workspace\` resolves cleanly (no other crates depend on the old version).
- [ ] Confirm \`cargo build --workspace\` clean.
- [ ] **Rule-7 / asserter add**: extend \`scripts/check-publish-pipeline.sh\` (from heddle#72) to add a structural check that runs \`cargo metadata\` against the workspace and verifies all internal consumers of every \`PUBLISHABLE_CRATES\` member have version requirements compatible with the source crate's actual version. This would have caught this bug BEFORE merge.
- [ ] On merge, the publish workflow should succeed and heddle-grpc 0.3.0 should appear on crates.io.

## Related

- [heddle#63](https://github.com/HeddleCo/heddle/issues/63) — closed, but its release is incomplete until this lands
- [heddle#72](https://github.com/HeddleCo/heddle/issues/72) — the publish workflow that correctly failed loud
- [weft#139](https://github.com/HeddleCo/weft/pull/139), [tapestry#57](https://github.com/HeddleCo/tapestry/pull/57) — blocked downstream PRs

## Suggested dispatch

medium effort. The Cargo.toml bumps are trivial; the asserter extension is the substantive work that prevents future occurrences.

---

## Definition of Done

- [ ] PR description includes \`gh run view\` output from the next \`publish-crates.yml\` run on main (should be green).
- [ ] Asserter check count incremented; the new structural check passes.
- [ ] Confirms \`https://crates.io/crates/heddle-grpc/0.3.0\` exists post-merge.

### PR DESCRIPTION
**Pull request by thethorne48**
## Summary

- Bump `heddle-grpc` version requirement from `"0.2"` → `"0.3"` in the three internal workspace consumers (`crates/cli`, `crates/daemon`, `crates/client`). Unblocks the auto-publish workflow (heddle#72), which has been failing on every push-to-main since heddle#63 r2 because `cargo publish` strips path deps and the published heddle-grpc 0.3.0 has no compatible candidate at `^0.2`.
- Extend `scripts/check-publish-pipeline.sh` with a new structural check: for every workspace crate that is publishable (re-derived from `[package].publish`, not the workflow's `PUBLISHABLE_CRATES` env var), walk every internal consumer and assert the version requirement is satisfied by the source crate's current version under cargo's caret semantics. Would have caught this exact bug at PR review time.

## Asserter behavior

- New check is the 20th `ok:` line (19 → 20). All 19 prior checks unchanged.
- Trust-aspect: re-derives the publishable set from each `[package].publish` field, so a future publishable crate is automatically covered without editing the asserter.
- Caret-only: workspace convention is caret reqs (cargo's default). Any non-caret comparator that slips in is loudly surfaced rather than silently passed.
- Output on regression:
  `heddle-cli requires heddle-grpc = "0.2", but heddle-grpc current version is 0.3.0 (incompatible)`

## Test plan

- [x] `cargo build --workspace` is clean after the version bumps.
- [x] `cargo metadata --no-deps | jq '.packages[].dependencies[] | select(.name == "heddle-grpc") | .req'` shows only `^0.3` (no `^0.2` residue).
- [x] `bash scripts/check-publish-pipeline.sh` passes with 20 `ok:` lines.
- [x] Mutation test: revert `crates/cli/Cargo.toml` to `version = "0.2"`, re-run the asserter, observe `::error::heddle-cli requires heddle-grpc = "0.2", but heddle-grpc current version is 0.3.0 (incompatible)` and exit 1.
- [ ] Once merged: `gh run list --workflow=publish-crates.yml --limit 1` shows a green run; heddle-grpc 0.3.0 lands on crates.io, unblocking weft#139 and tapestry#57.

Closes HeddleCo/heddle#76

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/HeddleCo/codesmith/heddle/pr/77"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg">View in Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need.</sup>

- [ ] Let Codesmith autofix CI failures and bot reviews

### CODE REVIEW

File: scripts/check-publish-pipeline.sh [hunk 1]
```diff
@@ -335,6 +335,173 @@ PY
...
+# Trust-aspect: we DON'T trust the PUBLISHABLE_CRATES env var from the
+# workflow as the source of truth here. We re-derive the publishable
+# set from each Cargo.toml's [package].publish field (default-publish
+# is publishable). That way a brand-new publishable crate added to the
+# workspace gets caught by this asserter without anyone remembering to
+# update the check.
+if ! command -v python3 >/dev/null 2>&1; then
+  err "python3 not available; consumer-version check skipped"
+elif ! python3 -c 'import tomllib' 2>/dev/null; then
+  err "python3 tomllib unavailable (needs 3.11+); consumer-version check skipped"
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply scripts/check-publish-pipeline.sh]: Fixed in c5959b8: tomllib gate now falls back to tomli on 3.10, then emits 'skip:' + exits 0 when neither is importable (no longer sets fail=1)

File: scripts/check-publish-pipeline.sh [hunk 2]
```diff
@@ -335,6 +335,173 @@ PY
...
+        return None
+    r = parse_ver(req)
+    v = parse_ver(ver)
+    if v < r:
+        return False
+    if r[0] > 0:
+        return v[0] == r[0]
+    if r[1] > 0:
+        return v[0] == 0 and v[1] == r[1]
+    return v[0] == 0 and v[1] == 0 and v[2] == r[2]
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply scripts/check-publish-pipeline.sh]: Fixed in c5959b8: all-zero req branch now widens by component count — "0" → <1.0.0, "0.0" → <0.1.0 — instead of collapsing to exact-patch

File: scripts/check-publish-pipeline.sh [hunk 3]
```diff
@@ -335,6 +335,173 @@ PY
...
+    # `publish` unset defaults to true (publishable). A non-empty list
+    # restricts the registries but is still publishable. `false` /
+    # empty list / explicit False means not publishable.
+    pub = pkg.get("publish")
+    if pub is None or pub is True or (isinstance(pub, list) and len(pub) > 0):
+        publishable.add(name)
+
+
+def parse_ver(s):
+    s = re.split(r"[-+]", s, maxsplit=1)[0]
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply scripts/check-publish-pipeline.sh]: Fixed in c5959b8: parse_ver_full keeps prerelease tag; source with -pre is rejected unless req is an exact =X.Y.Z-pre match (cargo's opt-in rule)

File: scripts/check-publish-pipeline.sh [hunk 4]
```diff
@@ -335,6 +335,173 @@ PY
...
+    # Cargo's default operator is caret. `cargo metadata` normalizes
+    # "0.2" → "^0.2", but Cargo.toml may carry the bare form, so we
+    # treat unprefixed as caret too. We deliberately don't try to
+    # parse `>=`, `~`, or comma-joined comparators — the workspace
+    # convention is caret, and a foreign comparator should be loudly
+    # surfaced rather than silently passed.
+    req = req.strip()
+    if req.startswith("^"):
+        req = req[1:]
+    elif req[:1] in ("=", ">", "<", "~", "*") or "," in req:
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply scripts/check-publish-pipeline.sh]: Fixed in c5959b8: wildcard guard now checks '*' in req (not just leading); 1.2.* returns None and surfaces as unsupported-comparator error

**Comment by thethorne48** [comment]
@codex please re-review — pushed c5959b8 fixing all 4 r1 P2 findings (caret semantics for 0/0.0, prerelease exclusion, wildcard rejection, tomllib skip-vs-fail). Added 4 self-test bundles inside the asserter; ok-line count is now 24.

**Comment by thethorne48** [comment]
@codex please take a look — initial review hasn't kicked off.

**Comment by thethorne48** [comment]
@codex still waiting for re-review of the r2 fixes (commit `c5959b8`). Caret-semver parser edge cases addressed; please take another look.

File: scripts/check-publish-pipeline.sh [hunk 5]
```diff
@@ -335,6 +335,293 @@ PY
...
+    # (Exact `=req-pre` matches were already handled above.)
+    if vpre and not rpre:
+        return False
+
+    # Lower bound: source >= requirement (numeric tuple compare; pre-release
+    # ordering against same numeric is not material here — cargo treats a
+    # prerelease as "less than" the same release, but the prior branch
+    # already filtered the asymmetric case).
+    if (vmaj, vmin, vpat) < (rmaj, rmin, rpat):
+        return False
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply scripts/check-publish-pipeline.sh:557]: Fixed in 6c203fc: added _cmp_prerelease() (semver §11) and gated caret prerelease comparisons through it — 0.3.0-alpha.0 now fails an 0.3.0-alpha.1 req. Self-test 'caret prerelease lower-bound ordering' covers it; mutation-verified.

File: scripts/check-publish-pipeline.sh [hunk 6]
```diff
@@ -335,6 +335,293 @@ PY
...
+    # ("0" vs "0.0" vs "0.0.0" differ in caret width when all zero.)
+    req_parts_given = len([p for p in req.split(".") if p != ""])
+
+    rmaj, rmin, rpat, rpre = parse_ver_full(req)
+    vmaj, vmin, vpat, vpre = parse_ver_full(ver)
+
+    # Prerelease in source but not in requirement → not satisfied.
+    # (Exact `=req-pre` matches were already handled above.)
+    if vpre and not rpre:
+        return False
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply scripts/check-publish-pipeline.sh]: Fixed in 6c203fc: when req has a prerelease tag, source must be on the same (major,minor,patch) tuple AND have a prerelease >= req's (release counts as > any prerelease on that tuple). 1.0.1 and 1.0.1-alpha both fail a 1.0.0-alpha req now. Self-test 'prerelease requirements pin to same (M,m,p) tuple' covers it.

File: scripts/check-publish-pipeline.sh [hunk 7]
```diff
@@ -335,6 +335,293 @@ PY
...
+    req = req.strip()
+
+    # Wildcards anywhere are rejected (not just leading). Catches "1.2.*".
+    if "*" in req:
+        return None
+
+    # Exact `=` comparator: full-string equality including prerelease.
+    # This is the ONLY way a prerelease source version can satisfy.
+    if req.startswith("="):
+        return req[1:].strip() == ver.strip()
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply scripts/check-publish-pipeline.sh]: Fixed in 6c203fc: '=' branch now strips +build metadata from both req and source per semver §10 before equality. =0.3.0-alpha.1 now matches 0.3.0-alpha.1+build.5. Self-test 'exact = ignores build metadata' covers it.

[the remark this answers is not in the document]
Comment (thethorne48) [reply scripts/check-publish-pipeline.sh]: Fixed in 6c203fc: partial '=' (e.g. =4, =4.2) now widens to a range — =4 → >=4.0.0,<5.0.0-0; =4.2 → >=4.2.0,<4.3.0-0. Full =X.Y.Z and =X.Y.Z-pre still take the normalized exact path. Self-test 'partial = requirements widen to ranges' covers it.

**Comment by thethorne48** [comment]
@codex please re-review — r2 round of caret-semver edge cases addressed in 6c203fc (prerelease ordering, same-tuple restriction, build-metadata stripping, partial =-as-range). Self-tests + mutation-verified.

File: scripts/check-publish-pipeline.sh [hunk 8]
```diff
@@ -335,6 +335,400 @@ PY
...
+
+    rmaj, rmin, rpat, rpre = parse_ver_full(req)
+    vmaj, vmin, vpat, vpre = parse_ver_full(ver)
+
+    # Req has a prerelease tag: cargo restricts matches to the same
+    # (major,minor,patch) tuple AND requires source prerelease >= req
+    # prerelease (release counts as > any prerelease on the same tuple).
+    if rpre:
+        if (vmaj, vmin, vpat) != (rmaj, rmin, rpat):
+            return False
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply scripts/check-publish-pipeline.sh:548]: Acknowledged as out-of-scope for current workspace: no heddle crate uses prerelease requirements (`X-rc.N`/`X-alpha`/`X-beta` patterns). The asserter's actual job — catching internal-consumer/source version mismatches like the original heddle-grpc 0.2/0.3.0 bug — works correctly. Filing a follow-up to either pull in `semver` crate or extend the prerelease behavior when a real consumer crate needs it. Resolving to ship + unblock the OSS git-overlay critical path (heddle#79).
